### ISSUE TITLE
Supporting manifest mode

### ISSUE THREAD

**Issue by MrMuetze** [mcgoo/vcpkg-rs#41]
Hey everyone 👋 
I was looking into vcpkg and how it could work with cargo and Rust and found this crate. I like how easily the integration works with the `build.rs` file, but since I would like to use the "version" feature of vcpkg. Yet versions are only supported in the manifest mode in which dependencies are read from the manifest file and then being installed in the project directory.

I had a look at the code and noticed that, to support manifest mode, the paths like e.g. `status_path`, `lib_path`, `bin_path` etc. need to be changed. Hardcoding a different root directory seemed to easily allow me to use this crate even in manifest mode.

Now I would like to discuss how to go about this. In manifest mode, a `vcpkg.json` file exists in the root of the Rust project. Would it make sense to check the existence of such a file and then automatically make the adjustments to the paths. Or should the user set an environment variable to let the crate now that it should look for packages in the project directory first? Or maybe it should always check the project directory first?

I'm hoping that we can make this crate even better!

Best regards,
Björn

**Comment by MrMuetze** [comment]
An example for a very simple approach can be seen here: https://github.com/GiGainfosystems/vcpkg-rs/commit/ccdf745011cf16512440140eb2d529e06b361d1a

Again, I do not know if that is a good approach, but it seems to work for the small examples that I've been trying.

**Comment by waych** [comment]
I don't know much about vcpkg's manifest mode, and haven't used it before, but I just read the online docs for it and I think I understand what you are trying to do.

I think it helps to differentiate the purpose and function of the `vcpkg-rs` crate vs `cargo-vcpkg`.  The `vcpkg-rs` crate simply reads info and files from an already installed vcpkg setup, whereas the `cargo-vcpkg` tool extends the `cargo` command to ease creating an installed vcpkg setup.

It sounds like the mode of operation you are using is to do the `vcpkg install` directly rather than via `cargo vcpkg install`, and then using `vcpkg-rs` to pick up those dependencies.  In that case I think something like an environment variable that defaults to ON according to presence of $(CARGO_MANIFEST_DIR)/vcpkg.json is probably okay (mimicking what VCPKG_MANIFEST_MODE does), so that by default it does the right thing (find the installed files in the new path location) but the behavior can be controlled.

Regarding `cargo-vcpkg`, it feels like some functionality overlaps between what vcpkg manifest mode does and the intent of cargo-vcpkg's metadata in Cargo.toml.  An potential path could have the metadata extended to cover versioning and have that generate a vcpkg.json on the backend, but that feels unnecessary if `vcpkg.json` is the upstream canonical way to express this.  In either case, it could use the `--x-install-root` to still keep the installed tree inside `./target` which seems desirable but not critical.

**Comment by MrMuetze** [comment]
You are correct. I haven't used `cargo vcpkg build` so far because it doesn't support pinning versions (yet). That's why I opted for the manifest mode of vcpkg directly. I think it's worth to update `vcpkg-rs` to support this mode as well and I think I will open a PR including your suggestion to automatically set this environment variable.

I have also just noticed that `cargo-vcpkg` fails to install as soon as a `vcpkg.json` file is present. I've opened an issue over there (https://github.com/mcgoo/cargo-vcpkg/issues/11). Also I agree with you that it makes sense to stay true to the specific purposes of both of these tools, but I don't know how deeply `cargo vcpkg` is integrated into `vcpkg-rs`, so it's hard for me to say much about this on my own. Some discussion might be in order for that.

**Comment by Be-ing** [comment]
I just did a quick experiment with vcpkg manifest mode running:
```
RUSTFLAGS="-L /path/to/vcpkg_installed/lib" PKG_CONFIG_PATH="/path/to/vcpkg_installed/lib/pkgconfig" cargo build
```
and that seems to works fine without a need for this crate nor `cargo vcpkg`.

**Comment by ras0219-msft** [comment]
From a conversation over on `system-deps` [1] https://github.com/gdesmott/system-deps/issues/8, they would like to use this package to improve their search paths for pkg-config to include vcpkg.

Of course, manifest mode is very important for this as well :)

**Comment by MrMuetze** [comment]
Just FYI I still have an eye on this but I'm currently enjoying my holidays so it'll take a bit until I'll respond more thoroughly 👍

**Comment by MrMuetze** [comment]
> I just did a quick experiment with vcpkg manifest mode running:
> 
> ```
> RUSTFLAGS="-L /path/to/vcpkg_installed/lib" PKG_CONFIG_PATH="/path/to/vcpkg_installed/lib/pkgconfig" cargo build
> ```
> 
> and that seems to works fine without a need for this crate nor `cargo vcpkg`.

Sorry for taking a while to come back to this. Do I understand correctly that these flags would have to be set for each individual library/dependency? If that is the case then I don't think that this crate or `cargo vcpkg` is superfluous.

**Comment by ras0219-msft** [comment]
You should only need to set `RUSTFLAGS="-L /path/to/vcpkg_installed/lib" PKG_CONFIG_PATH="/path/to/vcpkg_installed/lib/pkgconfig"` once. Because vcpkg puts all packages into a common sysroot, a single entry should be sufficient for pkg-config.

However, this crate provides integration with packages that don't provide .pc files. I think it could also be useful for a crate to mediate knowledge of a vcpkg instance for other packages; for example, would it be possible to automatically add the `RUSTFLAGS="-L /path/to/vcpkg_installed/lib" PKG_CONFIG_PATH="/path/to/vcpkg_installed/lib/pkgconfig"` based on knowledge of the instance?

**Comment by Diggsey** [comment]
Any update on this? The fork works great for me so would be nice to see this functionality upstreamed!

**Comment by heaths** [comment]
It seems the code already in the default branch is close. You do have code to check a `$VCPKG_INSTALLED_ROOT`, defaulting to "vcpkg_installed" in the working directory if not set. That's for manifest mode. But using procmon, I found that's short-circuited by a check for `status` in `$VCPKG_ROOT` which, for manifest mode, won't exist (assuming the only run on that host machine). Manifest mode solves a lot of problems and would seem to be the recommendation, given they call the other mode "classic".
